### PR TITLE
Add REST API for managing and posting to circles

### PR DESCRIPTION
**Pull request by Gargron**
Circles are the conceptual opposite of lists. A list is a subdivision of your follows, a circle is a subdivision of your followers. Posting
to a circle means making content available to only some of your followers. Circles have been internally supported in Mastodon for
the purposes of federation since #8950, this adds the REST API necessary for making use of them in Mastodon itsef.

Side note: Regardless of whether or not #12385 will ever be implemented, this is a prerequisite for that kind of system. Technically people could manually manage paid tiers of content this way already...

### CODE REVIEW

**Comment by trwnh** [comment]
to avoid confusion, can they be called "audiences"?

**Comment by Gargron** [comment]
I won't deny that both "circles" (popularized by Google Plus) and "aspects" (popularized by Diaspora) are not self-explanatory terms, so an alternative might be necessary. On the other hand, they *did* get popularized already (to some extent). I'm not a native English speaker but "audiences" sounds odd to me when pluralized. Perhaps something with more words would be easier to understand? Follower group? Follower bracket? Borrowing from Patreon, follower tier? Audience tier? (That one might get confused with public/unlisted/private/etc though)

**Comment by trwnh** [comment]
"Create an audience" and "Post to this audience" sound clear to me in English. "Audiences" doesn't sound weird to me as a native English speaker.

File: app/controllers/api/v1/circles/accounts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    @accounts = load_accounts
+    render json: @accounts, each_serializer: REST::AccountSerializer
+  end
+
+  def create
+    ApplicationRecord.transaction do
+      circle_accounts.each do |account|
+        @circle.accounts << account
+      end
+    end
```
Comment (ThibG) [inline app/controllers/api/v1/circles/accounts_controller.rb:22]: Why put that into a transaction?

File: app/models/circle_account.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  validates :account_id, uniqueness: { scope: :circle_id }
+
+  before_validation :set_follow
+
+  private
+
+  def set_follow
+    self.follow = Follow.find_by!(target_account_id: circle.account_id, account_id: account.id)
+  end
```
Comment (ThibG) [inline app/models/circle_account.rb:27]: What's the purpose of this?

File: app/services/post_status_service.rb [hunk 3]
```diff
@@ -67,7 +70,7 @@ def process_status!
     end
 
     process_hashtags_service.call(@status)
-    process_mentions_service.call(@status)
+    process_mentions_service.call(@status, @circle)
```
Comment (ThibG) [inline app/services/post_status_service.rb:73]: This is slightly awkward, at first I thought this didn't make sense, but given that we're implementing that through “silent mentions”… i guess this does make sense.

File: app/services/process_mentions_service.rb [hunk 4]
```diff
@@ -41,6 +42,12 @@ def call(status)
       "@#{mentioned_account.acct}"
     end
 
+    if circle.present?
+      circle.accounts.find_each do |target_account|
+        status.mentions.create(silent: true, account: target_account)
+      end
+    end
```
Comment (ThibG) [inline app/services/process_mentions_service.rb:49]: Won't this fail if someone from the circle is also explicitly mentioned? Because of the uniqueness constraint to mentions.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Thanks, that's something I've been wanting to implement for a while but couldn't get what your stand on this was, but also I have a few other concerns with.

In addition to the few remarks (including one which I think points to a bug), I have a few concerns:
- performances: storage-wise and federation-wise, posting a toot to a circle is functionally equivalent to writing a DM mentioning each of the circle's users, so i'm afraid this could be expensive with large circles
- UX: clients will see such messages as followers-only, not knowing they were intended for a subset of the followers, which may or may not be fine (e.g. you don't necessarily want recipients to know which circles they are part of; but on the other hand, as a poster you may want to know who you have posted a toot to)
- downgrade through reply: an extension of the issue above is that, since each client sees the toot as a followers-only toot, any reply to it will be sent as followers-only, effectively downgrading the privacy of the thread. This is similar to the issue we have with followers-only, except this time even a self-reply or a round-trip is, by default, a downgrade.
- compatibility with Pleroma: I think a while ago, Pleroma started implementing something similar. I think it worked in much the same way, but repurposed lists instead, and had special visibility to `list:id` instead of a new `circle` attribute
- API: which makes me think of, the circle being an extra parameter plays weridly with visibility scopes. furthermore, this doesn't seem to be part of the reblog endpoints either

File: app/models/circle_account.rb (hunk 2 above)
Comment (Gargron) [reply app/models/circle_account.rb:27]: If follow is deleted, this record will also be deleted

File: app/controllers/api/v1/circles/accounts_controller.rb (hunk 1 above)
Comment (Gargron) [reply app/controllers/api/v1/circles/accounts_controller.rb:22]: If one of the inserts fails, the client will see an error. Without transaction, it could lead to a partial success which could lead to confusion.

File: app/models/circle_account.rb (hunk 2 above)
Comment (ThibG) [reply app/models/circle_account.rb:27]: oh right, it's `not null` and `on_delete: cascade` in the SQL… though… why is it marked as optional?

File: app/controllers/api/v1/circles/accounts_controller.rb (hunk 1 above)
Comment (ThibG) [reply app/controllers/api/v1/circles/accounts_controller.rb:22]: Fair enough.

File: app/models/circle_account.rb (hunk 2 above)
Comment (Gargron) [reply app/models/circle_account.rb:27]: Hm... maybe that’s a mistake.

**Comment by noellabo** [comment]
CIrcleSerializer is missing. b83e87a655aa5907ad9d076fac1da9fcf9c37c99

File: app/services/process_mentions_service.rb (hunk 4 above)
Comment (noellabo) [reply app/services/process_mentions_service.rb:49]: Because it is not added to the mentions, it will not be created_notification().

```
mention = target_account.mentions.new(status: status, silent: true)
mentions << mention if mention.save
```

Comment (ThibG) [reply app/services/process_mentions_service.rb:49]: True. In addition, `create_notification` is kind of a misnomer because it handles both notifications and AP federation. In this case we probably do not want notifications (unless explicit mention), but we want to federate through AP.

**Comment by noellabo** [comment]
After thinking for a while, I put together my thoughts.

Mastodon, which does not support reply to limited visibility, has no choice but to allow downgrade to private. Pleroma recognizes it as direct visibility.

I think it is possible to build a mechanism to forward and redistribute a reply tree to limited status in Mastodon in the future, but it is a little complicated and it is difficult to make it compatible with Fediverse.

In that case, I think it is correct to handle the reply to limited status as direct visibility in Composer of WebUI and move to DM.

We changed the API to report limited visibility to clients as direct, and the limited flag compensates for this. Unlike direct, it will deliver to the home timeline instead of conversations.

In summary, the circle feature can deliver messages and content to a wider area than DM without revealing the destination. And you can receive the reaction directly by DM. It will be useful because it has different conveniences from DM.

On the other hand, the mechanism for sharing the member list and posting to all members can be covered by the group feature centered on the Group Actor. Group features and circle features are similar, but we should use them properly.

For circles, design decisions should be made about what to do and what not to do. I think keeping each function simple will be more widely accepted and used.

**Comment by ThibG** [comment]
Yeah, I basically see two ways to move forward with circles:
- handle replies as DMs, as you suggested: it's not super satisfactory because Circles are then completely one-way. Furthermore, current Mastodon versions will handle such toots as followers-only, not DMs. Finally, receiving a DM-like object without being in the explicit mentions reveals that you're in a Circle.
- handle circles as “anonymous groups”: followers-only and circle posts would be addressed to anonymous groups owned by the original author, and replies would be (by default) addressed to the group. Participants know who owns the group, and know that the group stay the same along a thread, but nobody but the group owner knows who's in the group. This means groups-based followers-only posts and circle-posts are indistinguishable, and that the audience is not accidentally widened (and when it is widened, people can easily notice). However, people replying cannot know who they are replying to and cannot restrict that to fewer people than who's in the original group (or at least, I do not know how we could do that). How, technically, to forward the toot to the group's members isn't technically obvious either (LDSigning would work, but it means the message would carry a proof of authorship which is not easily revokable, so that's something we want to avoid if possible).

**Comment by noellabo** [comment]
The circle_id is not recorded in the status, so the circle cannot be reselected during redraft or self reply.
I think it's better to save the circle_id in the owned status.

**Comment by noellabo** [comment]
Circle.title is only used visually, but I think it should be unique for identification purposes.

**Comment by Gargron** [comment]
Un-stale it, bot.
